README.md
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/4de000aaca86479ab0dcf13622117d31)](https://www.codacy.com/app/swzyocowboy/remap-indexed-keys-after-omit?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=iTonyYo/remap-indexed-keys-after-omit&amp;utm_campaign=Badge_Grade) [![Build Status](https://travis-ci.org/iTonyYo/remap-indexed-keys-after-omit.svg?branch=master)](https://travis-ci.org/iTonyYo/remap-indexed-keys-after-omit) [![Coverage Status](https://coveralls.io/repos/github/iTonyYo/remap-indexed-keys-after-omit/badge.svg?branch=master)](https://coveralls.io/github/iTonyYo/remap-indexed-keys-after-omit?branch=master) [![dependencies Status](https://david-dm.org/iTonyYo/remap-indexed-keys-after-omit/status.svg)](https://david-dm.org/iTonyYo/remap-indexed-keys-after-omit)

# `@oopsunome/remap-indexed-keys-after-omit`

针对键为 `index-x` 或 `x-index-y` 类似模式的被索引对象，如：

```javascript
{
  '0-0': 'one',
  '1-0': 'two',
  '2-0': 'three',
}
```

可在移除某个键值对后，顺序重新索引所有键。

[`@oopsunome/remap-indexed-keys-after-omit`][@oopsunome/remap-indexed-keys-after-omit] **支持在浏览器以及 Node.js 环境下使用**。Node.js 方面，持续且仅支持最新 LTS 版本的。

## 目录
- [安装](#安装)
- [使用](#使用)
    - [`remapIndexedKeysAfterOmitSync(options)`](#remapindexedkeysafteromitsyncoptions)
    - [`remapIndexedKeysAfterOmit(options)`](#remapindexedkeysafteromitoptions)
- [参与开发](#参与开发)
- [贡献指南](#贡献指南)
- [证书](#证书)
- [待办](#待办)

## 安装

```shell
# 使用 NPM
$ npm i @oopsunome/remap-indexed-keys-after-omit lodash

# 使用 Yarn
$ yarn add @oopsunome/remap-indexed-keys-after-omit lodash

# 使用 PNPM
$ pnpm install @oopsunome/remap-indexed-keys-after-omit lodash
```

## 使用

#### `remapIndexedKeysAfterOmitSync(options)`

- `options` {Object}
  - `object` {Object} 从小到大索引的对象
  - `omitIndex` {Number} 待删除键值对的索引
  - `keyPattern` {String} 对象的键模式
  - `separator` {String} 分裂键的分隔符
  - `indexPlaceholderInKeyPattern` {String} 指定索引占位
  - `gap` {Number} 递减 / 递增幅度
- 返回: {Object} 被顺序重新索引所有键的对象

```javascript
import { remapIndexedKeysAfterOmitSync } from '@oopsunome/remap-indexed-keys-after-omit';

(function() {
  const rslt = remapIndexedKeysAfterOmitSync({
    object: {
      '0-0': 254,
      '1-0': 7892,
      '2-0': 67568,
      '3-0': 23,
      '4-0': 657,
      '5-0': 3245,
      '6-0': 789,
      '7-0': 234,
      '8-0': 456,
      '9-0': 675,
      '10-0': 596,
    },
    omitIndex: 9,
    keyPattern: 'index-x',
    separator: '-',
    indexPlaceholderInKeyPattern: 'index',
    gap: 1,
  });

  console.table(rslt);

  /**
   * 输出:
   * 
   * {
   *   '0-0': 254,
   *   '1-0': 7892,
   *   '2-0': 67568,
   *   '3-0': 23,
   *   '4-0': 657,
   *   '5-0': 3245,
   *   '6-0': 789,
   *   '7-0': 234,
   *   '8-0': 456,
   *   '9-0': 596,
   * }
   */
})();
```

#### `remapIndexedKeysAfterOmit(options)`

- `options` {Object}
  - `object` {Object} 从小到大索引的对象
  - `omitIndex` {Number} 待删除键值对的索引
  - `keyPattern` {String} 对象的键模式
  - `separator` {String} 分裂键的分隔符
  - `indexPlaceholderInKeyPattern` {String} 指定索引占位
  - `gap` {Number} 递减 / 递增幅度
- 返回: {Promise} 被顺序重新索引所有键的对象

```javascript
import { remapIndexedKeysAfterOmit } from '@oopsunome/remap-indexed-keys-after-omit';

(async () => {
  console.table(await remapIndexedKeysAfterOmit({
    object: {
      '0-0': 254,
      '1-0': 7892,
      '2-0': 67568,
      '3-0': 23,
      '4-0': 657,
      '5-0': 3245,
      '6-0': 789,
      '7-0': 234,
      '8-0': 456,
      '9-0': 675,
      '10-0': 596,
    },
    omitIndex: 9,
    keyPattern: 'index-x',
    separator: '-',
    indexPlaceholderInKeyPattern: 'index',
    gap: 1,
  }));

  /**
   * 输出:
   * 
   * {
   *   '0-0': 254,
   *   '1-0': 7892,
   *   '2-0': 67568,
   *   '3-0': 23,
   *   '4-0': 657,
   *   '5-0': 3245,
   *   '6-0': 789,
   *   '7-0': 234,
   *   '8-0': 456,
   *   '9-0': 596,
   * }
   */
})();
```

## 参与开发

**准备开发环境**

详细参见 [SETUP.md][SETUP.md]。

**安装依赖**

[`@oopsunome/remap-indexed-keys-after-omit`][@oopsunome/remap-indexed-keys-after-omit] 使用 [`Yarn`][Yarn] 包管理器，执行 `yarn install` 安装依赖。

**开始开发**

```shell
yarn start
```

**生产构建**

```shell
yarn build
```

**测试**

```shell
yarn test
```

## 贡献指南

仔细查阅 [CONTRIBUTING.md][贡献指南] 以了解详情。

## 证书

[`@oopsunome/manual-sort`][@oopsunome/manual-sort] 获得了 MIT 许可，仔细查阅 [LICENSE.md][证书] 以了解详情。

## 待办

- [ ] 支持操作 `从大到小索引` 的对象；
- [ ] 为所有的配置提供默认值；
- [X] 支持 `解构` 或 `直接引用` 2 种暴露接口的方式；
- [X] 提供同步、异步 2 种编程方式，异步编程采用 `Promise`；
- [X] 支持构建 `CommonJS`, `AMD`, `ESM`, `UMD` 模式的模块；
- [X] 不将 `lodash` 等库打包；
- [X] 功能测试；
- [X] 捆绑 [Git 倒钩][Git倒钩]；
- [X] ESlint 检测；
- [X] 自动化生成所有依赖的开源证书；
- [X] 使用 [David DM][DavidDM] 实现 `依赖是否最新` 检测；
- [X] 使用 [Travis CI][TravisCI] 实现持续集成；
- [X] 使用 [Coveralls][Coveralls] 可视化测试用例覆盖率；
- [X] 使用 [Codacy][Codacy] 实现代码质量检测；
- [X] 更新日志；
- [ ] 自动化发布；
- [ ] 在什么场景下使用 [`@oopsunome/remap-indexed-keys-after-omit`][@oopsunome/remap-indexed-keys-after-omit]?
- [ ] 编写 [开发环境指南][SETUP.md]；
- [ ] 文档：在哪里可以获得更多帮助？
- [ ] 文档：设计思想；
- [ ] 文档：维护策略；
- [ ] 文档：[编码风格指南][编码风格指南]；
- [ ] 编写 [发布流程指南][发布流程指南]；
- [ ] 编写 [Git 指南][Git指南]；
- [ ] 编写 [命名指南][命名指南]；
- [ ] 编写 [版本指南][版本指南]；
- [ ] 性能测试；


[编码风格指南]: #
[版本指南]: #
[命名指南]: #
[Git指南]: #
[发布流程指南]: #
[Git倒钩]: https://github.com/typicode/husky
[DavidDM]: https://david-dm.org/
[TravisCI]: https://travis-ci.org/
[Coveralls]: https://coveralls.io/
[Codacy]: https://www.codacy.com/
[@oopsunome/remap-indexed-keys-after-omit]: https://github.com/iTonyYo/remap-indexed-keys-after-omit
[SETUP.md]: #
[Yarn]: https://yarnpkg.com/zh-Hans/
[贡献指南]: https://github.com/iTonyYo/remap-indexed-keys-after-omit/blob/master/CONTRIBUTING.md
[证书]: https://github.com/iTonyYo/remap-indexed-keys-after-omit/blob/master/LICENSE.md


CHANGELOG.md
# [`@oopsunome/remap-indexed-keys-after-omit`](https://github.com/iTonyYo/remap-indexed-keys-after-omit) 更新日志

## [v1.0.3](https://github.com/iTonyYo/remap-indexed-keys-after-omit/compare/v1.0.2...v1.0.3) - 2019-04-12

#### 提交

- [[`3ecd9f0`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/3ecd9f0d65a8c073c0079e34b3d2be41c72912bf)] 工作流: :railway_track: 升级 Mocha 至 v6.x.x+
## [v1.0.2](https://github.com/iTonyYo/remap-indexed-keys-after-omit/compare/v1.0.1...v1.0.2) - 2019-04-12

#### 提交

- [[`2ca226c`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/2ca226c5ab2b06b175ebd9b2aa51c50f2b4ed253)] 依赖: :package: 升级依赖
- [[`6dd21b9`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/6dd21b9d0f298d1044d3e94442c3eacf6f9e5cab)] 文档: :memo: 将 Git 提交指南从 CONTRIBUTING.md 中分离到 `./doc` 目录下
- [[`41693da`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/41693da74ab1832c2323edc78abd130e3fdd1045)] 工作流: :railway_track: 更新 VS Code 编辑器配置
- [[`a24c5a6`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/a24c5a6ed2f1f7e16d2c0ca208d785c1876dc9d9)] 工程: :clapper: 新增若干待办标记 & 变更编辑器的颜色主题
- [[`732b905`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/732b905bb490ae89b4c425531d9c249d8679f199)] 工程: :arrow_up: 升级若干开发依赖
- [[`d42e309`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/d42e309b58fdb88696a826d54505a69db2451624)] 工程: :arrow_up: 升级开发依赖
- [[`31bccf2`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/31bccf22b259ebe06e83451a06aa2a72b87b5546)] 发版: :bookmark: v1.0.2
- [[`101e6a4`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/101e6a4d672f8f34c7890165795840df388c3261)] 工程: :clapper: 新增若干关键字
- [[`79d6ed0`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/79d6ed0187ce2f982926bd905d2ffd0cce185a90)] 文档: :memo: 提供专门针对 Github 的问题、混合请求消息模板
- [[`8c39024`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/8c39024e3c032f5806b823537c4d075388d5368a)] 工作流: :railway_track: 详细配置 `@babel/env`
- [[`a5eff66`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/a5eff6640291508a107d680b8c5bd1095810f77d)] 工作流: :railway_track: 使 NPM 输出最小日志
- [[`5ce3907`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/5ce39070837aa4016c58a927aa9c0fe83da6049a)] 工作流: :railway_track: 执行 `build` 指令时仅删除上次构建，不删除其它诸如证书列表等
- [[`7651bdb`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/7651bdb816bf64a34f681418b4f1f7c0b396935e)] 工作流: :railway_track: 增加若干提交主题
## [v1.0.1](https://github.com/iTonyYo/remap-indexed-keys-after-omit/compare/v1.0.0...v1.0.1) - 2019-01-09

#### 提交

- [[`33dfcda`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/33dfcda43a8eac0ac5632ebf90693c73e11b262d)] 文档: :memo: 更新文档
- [[`bdf7c5d`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/bdf7c5dcd3cbf07da95a952c7436978d87c1ff4e)] 工程: :clapper: 调整更新日志模板
## v1.0.0 - 2019-01-09

#### 提交

- [[`a7aed6b`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/a7aed6b3a5b6dd29183e70909671537f78bd4eb2)] 工程: :clapper: 更新依赖
- [[`20ae34e`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/20ae34e6855115ac0d96be73a565f4f69e316450)] 文档: :memo: 更新文档
- [[`f528ed7`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/f528ed71e823ee9ed5216ac6398351414d47dfab)] 工程：:rocket: 升级开发依赖
- [[`2d9999e`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/2d9999e870b9b9e672dfec2ab3518868c52bf065)] 工程：:tada: 捆绑 [Git 倒钩](https://github.com/typicode/husky)
- [[`3243de1`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/3243de1d8af23715d732870b1f74710830b4e99f)] 工程：:rocket: 042c2e8130181c46488cc4df122faee6f192b058
- [[`e2e1ce4`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/e2e1ce483c2e3ec37cc2f027a1edae7234c78acf)] 性能: :racehorse: 避免引入完整的 `lodash`
- [[`596081f`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/596081fe41d6a9d0414ee7d68f7cb79b7e6d22a2)] 工程：:rocket: 升级开发依赖
- [[`01b0655`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/01b06553a57aa7e06bf3a05ca2fedd23ca917d44)] 工程: :clapper: 更新自动化指令
- [[`4dbb2e8`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/4dbb2e8cd04e8a90970a2e68128a7cb2180d66c8)] 工程: :clapper: 更新指令、发布配置、依赖等
- [[`56c3691`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/56c3691155d256a23302a3d67bf69e26248fc4f2)] 工程: :clapper: 配置更新日志
- [[`3096b76`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/3096b76416bdd006f9b734ba7ede22e32b3ca9b5)] 工程: :clapper: :beginner: `./.github`
- [[`87867c7`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/87867c75153d90d37218f197edf8784dafdfc4c2)] 工程: :tada: 使用 [Travis CI](https://travis-ci.org/) 实现持续集成
- [[`accca07`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/accca076d6cbb26f186759f380d274980afb363a)] 文档：:memo: 移除无效字符
- [[`381b7cd`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/381b7cd8739625657673136017efe696b839bb1f)] 工程：:rocket: 更新 `eslint` 忽略规则
- [[`6e8da57`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/6e8da57faf551b8423fe0f2bb2de923ad0974833)] 工程：:tada: 使用 [David DM](https://david-dm.org/) 实现 `依赖是否最新` 检测
- [[`842e541`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/842e54139833a61356d7be62beb51928f58ce793)] 工程：:rocket: 移除 v7.10.1 版本的 Node.js 运行时
- [[`a4238d7`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/a4238d787fec0501e544f36d0df37394921d761a)] 工程：:rocket: 让 Travis CI 使用 NPM 而非 Yarn
- [[`fdc0492`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/fdc04924e84d252842dd8809c99c81e07cd597c1)] 工程: :clapper: 更新 NPM 配置
- [[`372e041`](https://github.com/iTonyYo/remap-indexed-keys-after-omit/commit/372e041d569b6e51b5537a01c66bb816ef3d9d6c)] 工程: :clapper: 更新 `mocha` 与 `babel` 混合使用的配置

dev.js
require('@babel/register')();

const remapIndexedKeysAfterOmitSync
  = require('./remap-indexed-keys-after-omit-sync');

const remapIndexedKeysAfterOmit
  = require('./remap-indexed-keys-after-omit');

function printSync() {
  console.log(`> sync`);

  console.log(remapIndexedKeysAfterOmitSync({
    object: {
      '0-0': 254,
      '1-0': 7892,
      '2-0': 67568,
      '3-0': 23,
      '4-0': 657,
      '5-0': 3245,
      '6-0': 789,
      '7-0': 234,
      '8-0': 456,
      '9-0': 675,
      '10-0': 596,
    },
    omitIndex: 9,
    indexInKeyPattern: 'index-x',
    separator: '-',
    indexPlaceholder: 'index',
    gap: 1
  }));
}

function printAsync() {
  console.log(`\n> async`);

  remapIndexedKeysAfterOmit({
    object: {
      '0-0': 254,
      '1-0': 7892,
      '2-0': 67568,
      '3-0': 23,
      '4-0': 657,
      '5-0': 3245,
      '6-0': 789,
      '7-0': 234,
      '8-0': 456,
      '9-0': 675,
      '10-0': 596,
    },
    omitIndex: 9,
    indexInKeyPattern: 'index-x',
    separator: '-',
    indexPlaceholder: 'index',
    gap: 1
  })
  .then((result) => {
    console.log(result);
  })
  .catch((err) => {
    throw err;
  });
}

module.exports = (function() {
  printSync();
  printAsync();
})();
